### PR TITLE
Aquafarm withdraw/deposit support

### PR DESCRIPTION
**Pull request by 0xscuba**
### Context

* Current withdraw/deposit does not support aquafarm (reward earning pool token). It only works with base pools (none reward earning)
* Implement logic to
  * Convert pool tokens to aquafarm tokens when depositing
  * Revert aquafarm tokens to pool tokens when withdrawing

### Notes

* Installed `@orca-so/aquafarm`
* Added a new `getFarm()` api
* Added a new `OrcaFarm` interface and implementation for interacting with farm: `getFarmBalance`, `getFarmSupply`,  `deposit`, `withdraw`
* Added new constants in `orcaFarmConfigs` conforming to `OrcaFarmParams`

### Usage

* OrcaPool Deposit -> buidl deposit pool txn -> if aquafarm exists, get ref of farm object -> build farm deposit txn -> run txn
* OrcaPool Withdraw -> if aquarfarm ref exists, get ref of farm & build withdraw txn, build pool withdraw txn -> run txn

### Tests

* Tested `deposit()` and verified that aquafarm tokens were received; also verified on the website.
* Tested `withdraw()` and verified that aquafarm tokens could be redeemed back to tokenA and tokenB.

### CODE REVIEW

**Review by odcheung — COMMENTED** [review]
I think it's better if we handle aquafarm as a separate concept - both in terms of the data structure in pools.ts & in the orca-pool. A pool can own a reference to a farm, but a farm can also be a standalone thing (ex. rewards is a farm of a token & double-dip is a farm of a farm). It will also help us prevent the weirdness of having an available harvest ftn in a pool that doesn't have a farm.

1. OrcaPool Deposit -> buidl deposit pool txn -> if aquafarm exists, get ref of farm object -> build farm deposit txn -> run txn
2. OrcaPool Withdraw -> if aquarfarm ref exists, get ref of farm & build withdraw txn, build pool withdraw txn -> run txn
3. User checks pool for aquafarm ref -> Aquafarm.getHarvestableAmount -> Aquafarm.Harvest

File: src/model/orca/pool/orca-pool.ts [hunk 1]
```diff
@@ -50,7 +55,10 @@ export class OrcaPoolImpl implements OrcaPool {
   }
 
   public async getLPBalance(owner: PublicKey): Promise<OrcaU64> {
-    const address = await deriveAssociatedTokenAddress(owner, this.poolParams.poolTokenMint);
+    // Use farm balance if aquafarm exists for this pool, default to pool balance
+    const mintAddress =
```
Comment (odcheung) [inline src/model/orca/pool/orca-pool.ts:59]: getLPBalance of a pool should return the lp balance of the pool token amount. If the user wants to get the # of farm token balance, they should do so in a separate farm class

File: src/model/orca/pool/orca-pool.ts [hunk 2]
```diff
@@ -260,14 +271,84 @@ export class OrcaPoolImpl implements OrcaPool {
...
       .addInstruction(resolveTokenAInstrucitons)
       .addInstruction(resolveTokenBInstrucitons)
       .addInstruction(resolvePoolTokenInstructions)
       .addInstruction(transferTokenAInstruction)
       .addInstruction(transferTokenBInstruction)
-      .addInstruction(depositInstruction)
-      .build();
+      .addInstruction(depositInstruction);
+
+    if (this.poolParams.aquafarmParams) {
```
Comment (odcheung) [inline src/model/orca/pool/orca-pool.ts:282]: recommend we separate this out to a farm class

File: src/public/utils/web3/instructions/pool-instructions.ts [hunk 3]
```diff
@@ -156,3 +157,91 @@ export const createWithdrawInstruction = async (
     signers: owner.signer ? [owner.signer] : [],
   };
 };
+
+export const createInitUserFarmInstruction = async (
```
Comment (odcheung) [inline src/public/utils/web3/instructions/pool-instructions.ts:161]: separate out to farm-instructions

**Review by odcheung — COMMENTED** [review]
overall looks good! my comment is just on having aquafarm as a separate public api

**Comment by 0xscuba** [comment]
@odcheung - thanks for the review! Agree, that having aquafarm as a separate api makes more sense, as it will be more extensible for the future (e.g. when it comes to double-dip). Pushed a new commit that extracts the farm logic and created a new farm api, along with new constants.

File: src/public/farms/types.ts [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+
+  /**
+   * Query the supply of farm tokens for this pool.
+   * @return Returns the supply of farm tokens for this pool.
+   */
+  getFarmSupply: () => Promise<OrcaU64>;
+
+  /**
+   * TODO add comments
+   */
```
Comment (odcheung) [inline src/public/farms/types.ts:21]: fill this in as this is public to users

File: src/model/orca/farm/orca-farm.ts [hunk 5]
```diff
@@ -0,0 +1,238 @@
...
+    const context = await this.connection.getTokenSupply(this.farmParams.farmTokenMint);
+
+    const amt = new u64(context.value.amount);
+
+    return OrcaU64.fromU64(amt, this.farmParams.poolTokenDecimals);
+  }
+
+  public async deposit(
+    owner: Keypair | PublicKey,
+    poolTokenAmount: Decimal | OrcaU64
```
Comment (odcheung) [inline src/model/orca/farm/orca-farm.ts:62]: rename to baseTokenAmount

**Review by odcheung — COMMENTED** [review]
some nits on the naming & TODOs but otherwise the logic looks great!

File: src/public/farms/config.ts [hunk 6]
```diff
@@ -0,0 +1,41 @@
+/**
+ * The following content is auto-generated.
+ */
+
+/**
+ * A list of supported Orca farms in this SDK.
+ */
+export enum OrcaFarmConfig {
```
[the remark this answers is not in the document]
Comment (0xscuba) [reply src/public/farms/config.ts:8]: Thanks for the heads up! We might have to change the api depending on what we do. I tried to preserve the api and one thing that seems to work is like this:

```typescript
const x = OrcaFarmConfig["2uVjAuRXavpM6h1scGQaxqb6HVaNRn6T2X7HHXTabz25"];
```

In this way, we keep `OrcaFarmConfig` and don't change api, but need to keep it as an `enum`.

**Review by odcheung — APPROVED** [review]

**Pull request merged**